tsdb/chunkenc/bstream.go

```diff
@@ -111,90 +109,127 @@ func (b *bstream) writeBits(u uint64, nbits int) {
+	if b.valid == 0 {
+		if !b.loadNextBuffer() {
 			return false, io.EOF
 		}
-		b.count = 8
 	}
 
-	d := (b.stream[0] << (8 - b.count)) & 0x80
-	b.count--
-	return d != 0, nil
-}
-
-func (b *bstream) ReadByte() (byte, error) {
-	return b.readByte()
+	return b.readBitFast()
 }
 
-func (b *bstream) readByte() (byte, error) {
-	if len(b.stream) == 0 {
-		return 0, io.EOF
+// readBitFast is like readBit but can return io.EOF if the internal buffer is empty.
+// If it returns io.EOF, the caller should retry reading bits calling readBit().
+func (b *bstreamReader) readBitFast() (bit, error) {
+	if b.valid == 0 {
+		return false, io.EOF
 	}
 
-	if b.count == 0 {
-		b.stream = b.stream[1:]
+	b.valid--
+	bitmask := uint64(1) << b.valid
+	return (b.buffer & bitmask) != 0, nil
+}
 
-		if len(b.stream) == 0 {
+func (b *bstreamReader) readBits(nbits uint8) (uint64, error) {
+	if b.valid == 0 {
+		if !b.loadNextBuffer() {
 			return 0, io.EOF
 		}
-		return b.stream[0], nil
 	}
 
-	if b.count == 8 {
-		b.count = 0
-		return b.stream[0], nil
+	if nbits <= b.valid {
+		return b.readBitsFast(nbits)
 	}
 
-	byt := b.stream[0] << (8 - b.count)
-	b.stream = b.stream[1:]
+	// We have to read all remaining valid bits from the current buffer and a part from the next one.
+	bitmask := (uint64(1) << b.valid) - 1
+	nbits -= b.valid
+	v := (b.buffer & bitmask) << nbits
+	b.valid = 0
 
-	if len(b.stream) == 0 {
+	if !b.loadNextBuffer() {
 		return 0, io.EOF
 	}
 
-	// We just advanced the stream and can assume the shift to be 0.
-	byt |= b.stream[0] >> b.count
+	bitmask = (uint64(1) << nbits) - 1
+	v = v | ((b.buffer >> (b.valid - nbits)) & bitmask)
+	b.valid -= nbits
 
-	return byt, nil
+	return v, nil
 }
 
-func (b *bstream) readBits(nbits int) (uint64, error) {
-	var u uint64
+// readBitsFast is like readBits but can return io.EOF if the internal buffer is empty.
+// If it returns io.EOF, the caller should retry reading bits calling readBits().
+func (b *bstreamReader) readBitsFast(nbits uint8) (uint64, error) {
+	if nbits > b.valid {
+		return 0, io.EOF
+	}
 
-	for nbits >= 8 {
-		byt, err := b.readByte()
-		if err != nil {
-			return 0, err
-		}
+	bitmask := (uint64(1) << nbits) - 1
+	b.valid -= nbits
 
-		u = (u << 8) | uint64(byt)
-		nbits -= 8
+	return (b.buffer >> b.valid) & bitmask, nil
+}
+
+func (b *bstreamReader) ReadByte() (byte, error) {
+	v, err := b.readBits(8)
+	if err != nil {
+		return 0, err
 	}
+	return byte(v), nil
+}
 
-	if nbits == 0 {
-		return u, nil
+func (b *bstreamReader) loadNextBuffer() bool {
+	if b.streamOffset >= len(b.stream) {
+		return false
 	}
 
-	if nbits > int(b.count) {
-		u = (u << uint(b.count)) | uint64((b.stream[0]<<(8-b.count))>>(8-b.count))
-		nbits -= int(b.count)
-		b.stream = b.stream[1:]
+	// Handle the most common case in a optimized way.
+	if b.streamOffset+8 <= len(b.stream) {
+		// This is ugly, but significantly faster.
+		b.buffer =
+			((uint64(b.stream[b.streamOffset])) << 56) |
```

Comment:
how you found this?? :scream:

Reply:
Benchmarking it. Few % iteration time reduction comes from this.

Reply:
We could use `binary.BigEndian.Uint64` :)